runtime/vm/resolvesupport.cpp

```diff
@@ -2100,61 +2240,81 @@ resolveConstantDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
-	if (vmThread->currentException != NULL) {
-		/* Already a pending exception */
-		methodHandle = NULL;
-	} else if (methodHandle == NULL) {
-		setCurrentExceptionUTF(vmThread, J9VMCONSTANTPOOL_JAVALANGNULLPOINTEREXCEPTION, NULL);
-	}
-
-	/* Only write the value in if its not null */
-	if (NULL != methodHandle) {
-		J9MemoryManagerFunctions *gcFuncs = vmThread->javaVM->memoryManagerFunctions;
-		methodHandle = gcFuncs->j9gc_objaccess_asConstantPoolObject(
-									vmThread,
-									methodHandle,
-									J9_GC_ALLOCATE_OBJECT_TENURED | J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE | J9_GC_ALLOCATE_OBJECT_HASHED);
-
-		if (NULL == methodHandle) {
-			setHeapOutOfMemoryError(vmThread);
-		} else {
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+		/* Check if an exception is already pending */
+		if (vmThread->currentException != NULL) {
+			/* Already a pending exception */
+			result = vmThread->currentException;
+		} else if (result == NULL) {
+			setCurrentExceptionUTF(vmThread, J9VMCONSTANTPOOL_JAVALANGNULLPOINTEREXCEPTION, NULL);
+		} else { /* (result != NULL) */
+			/* Only write the value in if it is not null */
+			J9MemoryManagerFunctions *gcFuncs = vmThread->javaVM->memoryManagerFunctions;
 			J9Class *j9class = J9_CLASS_FROM_CP(ramCP);
-			if (0 == gcFuncs->j9gc_objaccess_staticCompareAndSwapObject(vmThread, j9class, callSite, NULL, methodHandle)) {
+
+			/* Ensure that result array elements are written before the array reference is stored */
+			VM_AtomicSupport::writeBarrier();
+			if (0 == gcFuncs->j9gc_objaccess_staticCompareAndSwapObject(vmThread, j9class, callSite, NULL, result)) {
 				/* Another thread beat this thread to updating the call site, ensure both threads return the same method handle. */
-				methodHandle = *callSite;
+				result = *callSite;
 			}
 		}
-	}
-
-	return methodHandle;
+#else /* defined(J9VM_OPT_OPENJDK_METHODHANDLE) */
+		/* Check if an exception is already pending */
+		if (vmThread->currentException != NULL) {
+			/* Already a pending exception */
+			result = NULL;
+		} else if (result == NULL) {
+			setCurrentExceptionUTF(vmThread, J9VMCONSTANTPOOL_JAVALANGNULLPOINTEREXCEPTION, NULL);
+		} else { /* (result != NULL) */
+			/* Only write the value in if it is not null */
+			J9MemoryManagerFunctions *gcFuncs = vmThread->javaVM->memoryManagerFunctions;
+			result = gcFuncs->j9gc_objaccess_asConstantPoolObject(
+										vmThread,
+										result,
+										J9_GC_ALLOCATE_OBJECT_TENURED | J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE | J9_GC_ALLOCATE_OBJECT_HASHED);
+
+			if (NULL == result) {
+				setHeapOutOfMemoryError(vmThread);
+			} else {
+				J9Class *j9class = J9_CLASS_FROM_CP(ramCP);
+				if (0 == gcFuncs->j9gc_objaccess_staticCompareAndSwapObject(vmThread, j9class, callSite, NULL, result)) {
+					/* Another thread beat this thread to updating the call site, ensure both threads return the same method handle. */
+					result = *callSite;
+				}
+			}
+		}
```

Comment:
Is there a `}` missing here?  The indentation doesn't match the `if (NULL == result) {` statement.